modules/ximgproc/src/paillou_filter.cpp

```diff
@@ -0,0 +1,454 @@
+        double				a1=2*exp(-a)*cosh(w)-exp(-2*a)-1;
+        double				b2=exp(-2*a);
+
+        switch(img.depth()){
+        case CV_8U :
+        {
+            for (int j=range.start;j<range.end;j++)
+            {
+                // Formule 26 p194
+                unsigned char *c1 = (unsigned char*)img.ptr(0)+j;// c1 pointeur sur x(i-1)
+                f2 = ((float*)im1.ptr(0))+j;
+                double border=*c1;
+				yp[0] = *c1 ;
+				c1+=cols;
+                yp[1] = *c1 - b1*yp[0]-b2*border;
+                c1+=cols;
+                for (int i=2;i<rows;i++,c1+=cols)
+                    yp[i] = *c1-b1*yp[i-1]-b2*yp[i-2];
+                // Formule 27 p194
+                c1 = (unsigned char*)img.ptr(rows-1)+j;
+                border=*c1;
+				ym[rows - 1] = *c1;
+				c1 -= cols;
+				ym[rows-2] =*c1 - b1*ym[rows-1];
+                c1 -= cols;
+                for (int i=rows-3;i>=0;i--,c1-=cols)
+                    ym[i]=*c1-b1*ym[i+1]-b2*ym[i+2];
+                // Formule 25 p193
+                for (int i=0;i<rows;i++,f2+=cols)
+                    *f2 = (float)(a1*(ym[i]-yp[i]));
+            }
+        }
+            break;
+        case CV_16S :
+		{
+			for (int j = range.start; j<range.end; j++)
+			{
+				// Formule 26 p194
+				short *c1 = (short*)img.ptr(0) + j;// c1 pointeur sur x(i-1)
+				f2 = ((float*)im1.ptr(0)) + j;
+				double border = *c1;
+				yp[0] = *c1;
+				c1 += cols;
+				yp[1] = *c1 - b1*yp[0] - b2*border;
+				c1 += cols;
+				for (int i = 2; i<rows; i++, c1 += cols)
+					yp[i] = *c1 - b1*yp[i - 1] - b2*yp[i - 2];
+				// Formule 27 p194
+				c1 = (short*)img.ptr(rows - 1) + j;
+				border = *c1;
+				ym[rows - 1] = *c1;
+				c1 -= cols;
+				ym[rows - 2] = *c1 - b1*ym[rows - 1];
+				c1 -= cols;
+				for (int i = rows - 3; i >= 0; i--, c1 -= cols)
+					ym[i] = *c1 - b1*ym[i + 1] - b2*ym[i + 2];
+				// Formule 25 p193
+				for (int i = 0; i<rows; i++, f2 += cols)
+					*f2 = (float)(a1*(ym[i] - yp[i]));
+			}
+		}
+		break;
+		case CV_16U :
+        {
+			for (int j = range.start; j<range.end; j++)
+			{
+				// Formule 26 p194
+				unsigned short *c1 = (unsigned short*)img.ptr(0) + j;// c1 pointeur sur x(i-1)
+				f2 = ((float*)im1.ptr(0)) + j;
+				double border = *c1;
+				yp[0] = *c1;
+				c1 += cols;
+				yp[1] = *c1 - b1*yp[0] - b2*border;
+				c1 += cols;
+				for (int i = 2; i<rows; i++, c1 += cols)
+					yp[i] = *c1 - b1*yp[i - 1] - b2*yp[i - 2];
+				// Formule 27 p194
+				c1 = (unsigned short*)img.ptr(rows - 1) + j;
+				border = *c1;
+				ym[rows - 1] = *c1;
+				c1 -= cols;
+				ym[rows - 2] = *c1 - b1*ym[rows - 1];
+				c1 -= cols;
+				for (int i = rows - 3; i >= 0; i--, c1 -= cols)
+					ym[i] = *c1 - b1*ym[i + 1] - b2*ym[i + 2];
+				// Formule 25 p193
+				for (int i = 0; i<rows; i++, f2 += cols)
+					*f2 = (float)(a1*(ym[i] - yp[i]));
+			}
+		}
+            break;
+        case CV_32S :
+        case CV_32F :
+        case CV_64F :
```

Comment:
It seems that this function doesn't work with these types of Mat data. I think it will be nice if you add some asserts for these types. In current implementation users of your functionality may be surprised why nothing works with CV_32S, CV_32F or CV_64F